utility/typing.py

```diff
@@ -21,6 +36,233 @@ def __bool__(self) -> bool:
+        replace_globals: dict[str, Any] = ...,
+        replace_locals: dict[str, Any] = ...,
+    ) -> dict[str, Any]: ...
+
+
+def resolve_annotations(
+    thing: Any,
+    /,
+    *,
+    extend_globals: dict[str, Any] = MISSING,
+    extend_locals: dict[str, Any] = MISSING,
+    filter_mro: Callable[[type[Any]], bool] = MISSING,
+    replace_globals: dict[str, Any] = MISSING,
+    replace_locals: dict[str, Any] = MISSING,
+    update_globals: dict[str, Any] = MISSING,
+    update_locals: dict[str, Any] = MISSING,
+) -> dict[str, Any]:
+    if getattr(thing, "__no_type_check__", None):
+        return dict()
+
+    annotations = dict()
+
+    if isinstance(thing, type):
+        for base in reversed(thing.__mro__):  # TODO: this can be optimized further by reading forward and never overwriting
+            if filter_mro is not MISSING and filter_mro(base):
+                continue
+
+            if getattr(base, "__no_type_check__", None):
+                continue  # TODO: is this behavior correct?
+
+            if "__annotations__" not in base.__dict__ or isinstance(base.__annotations__, types.GetSetDescriptorType):
+                continue
+
+            if replace_globals is not MISSING:
+                globals = dict(replace_globals)
+            else:
+                globals: dict[str, Any] = dict(getattr(sys.modules.get(base.__module__), "__dict__", dict()))
+
+                if extend_globals is not MISSING:
+                    for key, value in extend_globals.items():
+                        if key not in globals.keys():
+                            globals[key] = value
+
+                if update_globals is not MISSING:
+                    globals.update(update_globals)
+
+            if replace_locals is not MISSING:
+                locals = dict(replace_locals)
+            else:
+                locals = dict(base.__dict__)
+
+                if extend_locals is not MISSING:
+                    for key, value in extend_locals.items():
+                        if key not in locals.keys():
+                            locals[key] = value
+
+                if update_locals is not MISSING:
+                    locals.update(update_locals)
+
+            for name, annotation in base.__annotations__.items():
+                annotations[name] = resolve_annotation(annotation, globals=globals, locals=locals)
+    else:
+        if not hasattr(thing, "__annotations__"):
+            if isinstance(
+                thing,
+                (
+                    types.BuiltinFunctionType,
+                    types.FunctionType,
+                    types.MethodDescriptorType,
+                    types.MethodType,
+                    types.MethodWrapperType,
+                    types.ModuleType,
+                    types.WrapperDescriptorType,
+                )
+            ):
+                return dict()
+            else:
+                raise ValueError(f"thing {thing} is not annotated")
```

Comment:
Is not, or cannot be?
```suggestion
                raise ValueError(f"thing {thing} does not support annotations")
```

Reply:
I did this because you can inject annotations into a type that doesn't *support* them.